gcc/rust/Make-lang.in

```diff
@@ -461,3 +464,9 @@ rust/%.o: rust/checks/errors/borrowck/%.cc
 rust/%.o: rust/metadata/%.cc
 	$(COMPILE) $(RUST_CXXFLAGS) $(RUST_INCLUDES) $<
 	$(POSTCOMPILE)
+
+rust/libffi_polonius.a: \
+	rust/checks/errors/borrowck/ffi-polonius/Cargo.toml \
+	$(wildcard $(srcdir)/rust/checks/errors/borrowck/ffi-polonius/src/*)
+	cargo build --manifest-path $(srcdir)/rust/checks/errors/borrowck/ffi-polonius/Cargo.toml --release --target-dir rust/ffi-polonius
```

Comment:
I assume that requiring `cargo` as part of GCC/Rust build process is not exactly what upstream GCC was looking for...  🙃

Reply:
So, there is not simply a shared library to link against, or executable to call (at `GCC/Rust` invocation time)?

Reply:
That said, `cargo` is "just" a build tool -- so assumingly we can get FFI-Polonius built in another "standard" (for GCC practices) way?

Reply:
We would need to compile polonius with gccrs, which probably won't be possible for a long time.

Reply:
We could probably use direct `rustc` invocation instead of using `cargo`. The current `polonius-engine` only has 3 deps:
```
polonius-engine v0.13.
├── datafrog v2.0.1
├── log v0.4.14
│   └── cfg-if v1.0.0
└── rustc-hash v1.1.0
```
but this doesn't scale very well if more dependencies are added. Would having a dep on `rustc` to build gccrs something we want in GCC until gccrs can build polonius? If not, then this can stay in gccrs tree (for a long time...).

Is there already a plan for this?

Reply:
> I assume that requiring `cargo` as part of GCC/Rust build process is not exactly what upstream GCC was looking for... 🙃

Yeah, I mentioned that to @CohenArthur and he said we don't care, we need it. It is temporary until we can build Polonius ourselves (so like mid-term "temporary").

Reply:
> We could probably use direct `rustc` invocation instead of using `cargo`. The current `polonius-engine` only has 3 deps:
> 
> ```
> polonius-engine v0.13.
> ├── datafrog v2.0.1
> ├── log v0.4.14
> │   └── cfg-if v1.0.0
> └── rustc-hash v1.1.0
> ```
> 
> but this doesn't scale very well if more dependencies are added. Would having a dep on `rustc` to build gccrs something we want in GCC until gccrs can build polonius? If not, then this can stay in gccrs tree (for a long time...).
> 
> Is there already a plan for this?

Do you really think using plain `rustc` is better for the gcc folks?  (I would feel somewhat ashamed to build Rust with make 😢, but it is not really important here.)

Reply:
Have no idea. I feel like this change won't hit upstream gcc until...some time. So whatever is best for you, but this needs to not get in the way of upstreaming the rest of the compiler. If that's the plan (we commit things we know are not fit for upstream gcc), then you should probably have separate commits for changes that will be upstreamed and the changes that we know won't ever reach gcc in this state.

If you start requiring `rustc`/`cargo`, you'll break all platforms that are not supported by `rustc` (exit `sh-*`, `s390x-*`, ...)

Reply:
I clearly wasn't really awake yet, when posting my comments yesterday Monday morning...  

> requiring cargo in not really worse then rustc

Point taken.  ;-)

> compile polonius with gccrs

That also won't solve all problems immediately, given that we need Polonius compiled for the _host_ architecture (in order to link it into GCC), and _build_, _host_, _target_ architectures may all be different.  So we cannot rely on being able to use the GCC/Rust being built (for the _target_ architecture) for building any Rust code for _host_ usage.  (Building for _target_ is fine, of course, like other GCC target libraries.)

A multi-stage build process would resolve that issue; either manual or even scripted into the GCC's top-level build system (non-trivial).

My original assumption was:

> So, there is not simply a shared library to link against, or executable to call (at GCC/Rust invocation time)?

I agree to what Marc just said:

> whatever is best for you, but this needs to not get in the way of upstreaming the rest of the compiler

..., so we have to make use of Polonius conditional on its availability; and otherwise continue to require `-frust-incomplete-and-experimental-compiler-do-not-use` being specified, for instance.

Such a GCC/Rust mode generally will be necessary to continue supporting GCC/Rust in configurations where standard Rust is not available, and thus Polonius cannot be built at all, currently.  (Standard bootstrapping problem.)  ..., unless Polonius can be cross-compiled for the respective _host_ architecture, of course.

Reply:
> > compile polonius with gccrs
> 
> That also won't solve all problems immediately, given that we need Polonius compiled for the _host_ architecture (in order to link it into GCC), and _build_, _host_, _target_ architectures may all be different. So we cannot rely on being able to use the GCC/Rust being built (for the _target_ architecture) for building any Rust code for _host_ usage. (Building for _target_ is fine, of course, like other GCC target libraries.)  

Kind of similar issue with proc macros that need to be compiled for host (but not only during the build of gcc)

> If you start requiring `rustc`/`cargo`, you'll break all platforms that are not supported by `rustc` (exit `sh-*`, `s390x-*`, ...)

Not 100% correct, only where `rustc` is not available for host, not target (e.g. `sh` is usually cross compiled, so `rustc` doesn't need to be available for the `sh` target)

Reply:
Could polonius be loaded as a shared library by the compiler? The actual implementation, as opposed to the bindings, could be kept out-of-tree for now.